T23314: Add initial implementation AnimationsDbus library

smspillaz opened the pull request:
This PR adds an initial implementation of the server side of the AnimationsDbus library as described in the attached ticket.

The server side library is designed to be used like GNOME-Shell
or other compositors which implement surface level animations.

The server uses animations_dbus_server_new_async() to asynchronously
create a new AnimationsDbusServer object and corresponding
AnimationsDbusConnectionManager which will be exported on the bus. One
of the construct properties that the server must provide is an implementation
of a AnimationsDbusServerEffectFactory which is responsible for constructing
the effect "metaclass" (AnimationsDbusServerEffectBridge) on the server side.

On each surface that the server is managing, the server calls
animations_dbus_server_register_surface along with an
implementation of a AnimationsDbusServerSurfaceBridge, which
is an interface providing properties about the surface that
would be interesting to internal clients. Calling register_surface
will export an AnimationsDbusSurface object on the bus for clients
to interact with.

The AnimationsDbusServerSurfaceBridge implementation should have
an attach_effect method which takes an AnimationsDbusServerEffectBridge
and determines if it can be attached to the surface on that event. If
it can be, attach_effect should return an AnimationsDbusServerSurfaceAttachedEffect
implementation, behind which will be private data for that animation
(for instance, progress, spring state) that exists for that animation
on the given surface. It is the server's responsibility to actually
perform the animation using that private data and the settings for that
animation exist on the attached AnimationsDbusServerEffectBridge.

Clients create and attach animations by first registering themselves
using the RegisterClient method on the exported AnimationsDbusConnectionManager
which takes note of their bus name and returns an object path to an
AnimationsDbusAnimationManager (specific to that client), which they can
call CreateAnimation on to create new animation metaclasses. Clients can
then call the AttachEffect method on the exported AnimationsDbusSurface
to attach their effects to the surface. Once the bus name owned by the
client disappears from the bus, all the client specific state will
be automatically cleaned up (eg, animations are detached, private data
for attached animations is destroyed).

Surfaces can have multiple animations attached to a single event. In
case multiple animations are attached, the highest priority attached one
is selected. If the highest priority animation is destroyed, it is removed
from the list and the next highest priority animation will be used. In
general, the latest attached effect by a client has the highest priority.

Both external clients and the server can register animations with
the server (the server would create a new AnimationManager object
using the animation_dbus_server_create_animation_manager method
on the AnimationsDbusServer). When the server attaches an animation
to an AnimationsDbusServerSurface, it uses
the animations_dbus_server_surface_attach_animation_effect_with_server_priority
which gives that attached effect the lowest priority (for instance, if
an external client has already attached a temporary effect to a surface,
that effect has a higher priority than the one attached by the server). This
allows the server to, for instance, globally attach effects to all surfaces
as the "default", but with the ability for external clients on the bus
to override that configuration temporarily.

https://phabricator.endlessm.com/T23314

Commit:
data: Add Libanimation DBus specification

Commit:
animations-dbus: Add client library

The client library more or less thinly wraps the server library
and supports both async and sync operations. The sync operations
wrap the async operations in a pushed main context.

The benefit of using the client library is that there isn't the
extra step invovled in getting the object path and then creating
a proxy for the object, it can be done in one step.

wjt commented on animations-dbus/animations-dbus-server-animation-manager.h:
```diff
@@ -0,0 +1,53 @@
+/* Copyright 2018 Endless Mobile, Inc.
+ *
+ * animations-dbus is free software: you can redistribute it and/or
```
Is this library "animations-dbus" (as here) or "libanimation-dbus" (as in the next paragraph)?

smspillaz replied:
It should be `libanimation-dbus` ... looks like I'll need to run another regex.

smspillaz replied:
This should be fixed now ... not sure why its still showing as up-to-date

Commit:
examples: Add AnimationsDbus client example Javascript file

This shows how the client library can be used to iterate
through the available surfaces and attach animations to windows
(both async and sync).

smspillaz commented:
I should note that there's a couple of warnings here in the build due (a) `gdbus-codegen` generated code triggering a bunch of warnings about unused variables that Meson's default warning set triggers and (b) the fact that I needed to use a few typedefs to break `#include` cycles, which Meson's default warning set doesn't like either.

I'm not sure what can be done about (a), but if anyone has any ideas on how to deal with (b) properly (surely it can't be that hard), it would be appreciated.

smspillaz commented:
(I just saw now that automatic detaching of effects on client close doesn't seem to work. I'll fix that tonight).

joaquimrocha commented on animations-dbus/meson.build:
```diff
@@ -0,0 +1,42 @@
+    version_h
+]
+private_headers = [
+]
```
Nitpick: I'd but these on the same line as the opening one: `private_headers = []` seems clearer to me, but it's a very subjective thing.

smspillaz replied:
Fixed.

joaquimrocha commented on animations-dbus/meson.build:
```diff
@@ -0,0 +1,42 @@
+        '-DCOMPILING_DMODEL'],
+    dependencies: [gio, gio_unix, glib, gobject],
+    include_directories: include, install: true,
+    soversion: api_version, version: libtool_version)
```
Nitpick: I think it'd read better if there was one variable per line.

joaquimrocha replied:
Also, do you need to add the API version to the lib's name?
And I'd use `shared_library` here as it seems to be the one with the arguments you have: http://mesonbuild.com/Reference-manual.html#shared_library

smspillaz replied:
AIUI, the soversion should be sufficient, at least to handle the parallel-installability problem?

I also prefer a one-variable-per-line style, but I believe this is the default style for @endlessm projects (at least, according to @ptomato ?).

I switched this to use shared_library in the meantime.

joaquimrocha commented on animations-dbus/meson.build:
```diff
@@ -0,0 +1,42 @@
+
+introspection_sources = [
+    sources,
+    meson.build_root() + '/animations-dbus-objects.h',
```
Use `join_paths` please:
http://mesonbuild.com/Reference-manual.html#join_paths
Also, usually you don't need to add the headers to the source list if they're included in the actual source files. I prefer to omit them personally. But it's up to you.

smspillaz replied:
Handy, didn't know about that one.

I think including the headers is necessary to make introspection work.

joaquimrocha commented on animations-dbus/meson.build:
```diff
@@ -0,0 +1,42 @@
+    identifier_prefix: 'AnimationsDbus', include_directories: include,
+    includes: ['Gio-2.0', 'GLib-2.0', 'GObject-2.0'],
+    install: true, namespace: namespace_name, nsversion: api_version,
+    sources: introspection_sources, symbol_prefix: 'animations_dbus')
```
Nitpick: Again, one arg per line would read so much better IMO.

smspillaz replied:
I agree, though, see above.

joaquimrocha commented on meson.build:
```diff
@@ -0,0 +1,40 @@
+subdir('data')
+subdir('animations-dbus')
+
+requires = ['gio-2.0', 'gio-unix-2.0', 'glib-2.0', 'gobject-2.0']
```
You can actually use the dependency objects declared above, so you don't repeat the names here.

smspillaz replied:
Handy, fixed.

joaquimrocha commented on animations-dbus/animations-dbus-errors.c:
```diff
@@ -0,0 +1,46 @@
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with libanimation-dbus.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
```
Add yourself as the author? Credit where credit is due.
And the same in other occasions.

smspillaz replied:
;-)

joaquimrocha commented on animations-dbus/animations-dbus-server-animation-manager.h:
```diff
@@ -0,0 +1,53 @@
+#define ANIMATIONS_DBUS_TYPE_SERVER_ANIMATION_MANAGER animations_dbus_server_animation_manager_get_type ()
+G_DECLARE_FINAL_TYPE (AnimationsDbusServerAnimationManager, animations_dbus_server_animation_manager, ANIMATIONS_DBUS, SERVER_ANIMATION_MANAGER, AnimationsDbusAnimationManagerSkeleton)
+
+typedef struct _AnimationsDbusServer AnimationsDbusServer;
```
This and some other places generate a warning because of redefinition.
I think you're trying to simulate a forward declaration, but I guess the way to go is to have a 3rd file `animation-dbus-server-type.h` that you can include in both? Analogously for other similar warnings?

smspillaz replied:
Yeah, I think that would solve the problem. Thanks for the tip!

joaquimrocha commented on animations-dbus/animations-dbus-server-animation-manager.c:
```diff
@@ -0,0 +1,382 @@
+typedef struct _AnimationsDbusServerAnimationManagerPrivate
+{
+  GDBusConnection                *connection;
+  AnimationsDbusServer           *server;
```
These are not aligned with the one below. And IMO they shouldn't even be column aligned at all, to avoid cases like this where an addition makes all the alignment outdated.

joaquimrocha commented on animations-dbus/animations-dbus-server-animation-manager.c:
```diff
@@ -0,0 +1,382 @@
+ *          @error set if one does not exist for that ID on this @animation_manager.
+ */
+AnimationsDbusServerEffect *
+animations_dbus_server_animation_manager_lookup_effect_by_id (AnimationsDbusServerAnimationManager  *server_animation_manager,
```
I am starting to wonder whether going for an abbreviation of the namespace prefix would be recommended. Maybe `AD` or `AnimDbus`?

smspillaz replied:
I started to think the same thing a day ago too ... though I think that could be much more churn... perhaps better to do it once we've wrapped up the other review comments.

joaquimrocha commented on animations-dbus/animations-dbus-server-animation-manager.c:
```diff
@@ -0,0 +1,382 @@
+ */
+AnimationsDbusServerEffect *
+animations_dbus_server_animation_manager_lookup_effect_by_id (AnimationsDbusServerAnimationManager  *server_animation_manager,
+                                                              unsigned int                           effect_id,
```
guint?

smspillaz replied:
AIUI the convention for new code is to avoid `g*` typedef'd types.

joaquimrocha commented on animations-dbus/animations-dbus-server-animation-manager.c:
```diff
@@ -0,0 +1,382 @@
+                                                              GError                               **error)
+{
+  AnimationsDbusServerAnimationManagerPrivate *priv =
+    animations_dbus_server_animation_manager_get_instance_private (server_animation_manager);
```
Now that's a long name :slightly_smiling_face:

smspillaz replied:
They get worse :(

joaquimrocha commented on animations-dbus/animations-dbus-server-animation-manager.c:
```diff
@@ -0,0 +1,382 @@
+                                                        const char                            *title,
+                                                        const char                            *name,
+                                                        GVariant                              *settings,
+                                                        GError                               **error)
```
Personal-nitpick: I really dislike columned aligned args...

smspillaz replied:
AIUI it is the standard coding style across endless C projects

joaquimrocha commented on animations-dbus/animations-dbus-server-animation-manager.c:
```diff
@@ -0,0 +1,382 @@
+                                       settings);
+  const char *animation_manager_object_path =
+    g_dbus_interface_skeleton_get_object_path (G_DBUS_INTERFACE_SKELETON (server_animation_manager));
+  unsigned int available_serial = priv->animation_effect_serial++;
```
guint?

joaquimrocha commented on animations-dbus/animations-dbus-server-animation-manager.c:
```diff
@@ -0,0 +1,382 @@
+                   ANIMATIONS_DBUS_ERROR,
+                   ANIMATIONS_DBUS_ERROR_INTERNAL_ERROR,
+                   "%s",
+                   message);
```
Why do you format a string as a variable that will have to be allocated and freed instead of formatting it in the `g_set_error` message?

smspillaz replied:
Yeah, we can probably do the formatting inside `g_set_error`. I must have forgotten about that. Fixed

joaquimrocha commented on animations-dbus/animations-dbus-server-animation-manager.c:
```diff
@@ -0,0 +1,382 @@
+                   ANIMATIONS_DBUS_ERROR_INTERNAL_ERROR,
+                   "%s",
+                   message);
+      return NULL;
```
I think you're leaking `animation_effect` here, right?

smspillaz replied:
Good catch. It should be an autoptr

joaquimrocha commented on animations-dbus/animations-dbus-server-animation-manager.c:
```diff
@@ -0,0 +1,382 @@
+  g_hash_table_insert (priv->animation_effects,
+                       GUINT_TO_POINTER (available_serial),
+                       g_object_ref (animation_effect));
+  return g_steal_pointer (&animation_effect);
```
You forgot to declare the `animation_effect` as an auto pointer.

smspillaz replied:
Ah, I should have read this comment here :)

joaquimrocha commented on animations-dbus/animations-dbus-server-animation-manager.c:
```diff
@@ -0,0 +1,382 @@
+      return TRUE;
+    }
+
+  animations_dbus_animation_manager_complete_create_animation_effect (animation_manager,
```
:scream:

smspillaz replied:
T_T

joaquimrocha commented on animations-dbus/animations-dbus-server-effect-bridge-interface.c:
```diff
@@ -0,0 +1,40 @@
+
+  AnimationsDbusServerEffectBridgeInterface *iface = ANIMATIONS_DBUS_SERVER_EFFECT_BRIDGE_GET_IFACE (self);
+
+  g_return_val_if_fail (iface->get_name != NULL, FALSE);
```
You should return `NULL`.

joaquimrocha commented on animations-dbus/animations-dbus-server-effect.c:
```diff
@@ -0,0 +1,491 @@
+                           key,
+                           printed_value,
+                           animations_dbus_server_effect_bridge_get_name (priv->effect_bridge));
+                continue;
```
Isn't it necessary to free the key and value when it fails too? (i.e. remove the `continue`?)

smspillaz replied:
It is - good catch.

smspillaz replied:
In fact, we can remove the `continue` here since the last two lines just free the key and value anyway.

joaquimrocha commented on animations-dbus/animations-dbus-server-surface-bridge-interface.c:
```diff
@@ -0,0 +1,105 @@
+                                                     GError                            **error)
+{
+  g_return_val_if_fail (ANIMATIONS_DBUS_IS_SERVER_SURFACE_BRIDGE (self), NULL);
+  g_return_val_if_fail (error == NULL || *error == NULL, NULL);
```
Why do you force passing a `GError`?

smspillaz replied:
The error param can be `NULL`, just that if it is passed, it needs to be unset.

joaquimrocha commented on animations-dbus/animations-dbus-server-surface.c:
```diff
@@ -0,0 +1,691 @@
+{
+  AnimationsDbusServerSurfacePrivate *priv = animations_dbus_server_surface_get_instance_private (server_surface);
+
+  g_assert (push_func != NULL);
```
Should you use `g_return_if_fail`?

smspillaz replied:
I'm following chergert's guideline here, which is that internal API should use `g_assert` for preconditions and externally callable API should use `g_return...` checks.

joaquimrocha commented on animations-dbus/animations-dbus-server-surface.c:
```diff
@@ -0,0 +1,691 @@
+{
+  /* Other effects take priority over this one, since server effects
+   * can be overridden by the user. */
+  if (!animations_dbus_server_surface_attach_effect_with_queue_func (server_surface,
```
`return animations_dbus_server_...`?

smspillaz replied:
Also works, thanks!

joaquimrocha commented on animations-dbus/animations-dbus-server-surface.c:
```diff
@@ -0,0 +1,691 @@
+  g_autoptr(GError) local_error = NULL;
+
+  g_return_val_if_fail (out_animation_manager_id != NULL, FALSE);
+  g_return_val_if_fail (out_animation_effect_id != NULL, FALSE);
```
If you declare these at the top, then you save it from doing any work in splitting the string and counting the tokens...

smspillaz replied:
True, I'll move these

joaquimrocha commented on animations-dbus/animations-dbus-server-surface.c:
```diff
@@ -0,0 +1,691 @@
+                                                                     &local_error))
+    {
+      g_dbus_method_invocation_return_gerror (invocation,
+                                              g_steal_pointer (&local_error));
```
Should you return here?

smspillaz replied:
I should, yes - fixed.

joaquimrocha commented on animations-dbus/animations-dbus-server-surface.c:
```diff
@@ -0,0 +1,691 @@
+    {
+      g_dbus_method_invocation_return_gerror (invocation,
+                                              g_steal_pointer (&local_error));
+      return TRUE;
```
Should the returns here be `TRUE` on failure?

smspillaz replied:
I should, yes - good catch.

joaquimrocha commented on animations-dbus/animations-dbus-server-surface.c:
```diff
@@ -0,0 +1,691 @@
+  AnimationsDbusServerSurface *server_surface = ANIMATIONS_DBUS_SERVER_SURFACE (animatable_surface);
+  AnimationsDbusServerSurfacePrivate *priv = animations_dbus_server_surface_get_instance_private (server_surface);
+  unsigned int animation_manager_id = 0;
+  unsigned int animation_effect_id = 0;
```
guint?

smspillaz replied:
See above

joaquimrocha commented on animations-dbus/animations-dbus-server-surface.c:
```diff
@@ -0,0 +1,691 @@
+    {
+      g_dbus_method_invocation_return_gerror (invocation,
+                                              g_steal_pointer (&local_error));
+      return TRUE;
```
Should it return `FALSE`?

smspillaz replied:
AIUI, `TRUE` here means that the method call handler ran (regardless the error state is in the invocation return value). There's another place below where I return `FALSE` in error though, I'll fix that.

joaquimrocha commented on animations-dbus/animations-dbus-server.c:
```diff
@@ -0,0 +1,624 @@
+    animations_dbus_server_surface_new (priv->connection,
+                                        server,
+                                        bridge);
+  unsigned int allocated_id = priv->animatable_surface_serial++;
```
guint?

smspillaz replied:
See above

joaquimrocha commented on animations-dbus/animations-dbus-server.c:
```diff
@@ -0,0 +1,624 @@
+                                                     ANIMATIONS_DBUS_ERROR,
+                                                     ANIMATIONS_DBUS_ERROR_NAME_ALREADY_REGISTERED,
+                                                     message);
+      return TRUE;
```
Shouldn't it return `FALSE`?

smspillaz replied:
See above

joaquimrocha commented on animations-dbus/animations-dbus-server.c:
```diff
@@ -0,0 +1,624 @@
+                                                     ANIMATIONS_DBUS_ERROR,
+                                                     ANIMATIONS_DBUS_ERROR_INTERNAL_ERROR,
+                                                     message);
+      return TRUE;
```
Same here maybe?

smspillaz replied:
See above

cosimoc commented on examples/client.js:
```diff
@@ -0,0 +1,62 @@
+          let event = Object.keys(effects)[0];
+          let selectedEffectName = effects[event][0];
+
+          log(`Available effects are: ${JSON.stringify(effects, null, 2)}`);
```
I know this is just an example, but this line is redundant with the other one above.

smspillaz replied:
Not sure I understand what you mean here?

cosimoc commented on data/com.endlessm.Libanimation.xml:
```diff
@@ -0,0 +1,156 @@
+<node>
+  <interface name="com.endlessm.Libanimation.ConnectionManager">
+    <!--
+      RegisterClient() -> o: Create a new AnimationManager object on
```
`on the bus name` -> `for the bus name`

cosimoc commented on data/com.endlessm.Libanimation.xml:
```diff
@@ -0,0 +1,156 @@
+    <!--
+      RegisterClient() -> o: Create a new AnimationManager object on
+                             the bus name of the caller’s connection.
+                             The bus name should be the name owned by
```
`should be the name owned by the calling client and` -> `owned by the calling client`

smspillaz replied:
Thanks - this would have been a leftover from when we were explicitly passing the name ;-)

cosimoc commented on animations-dbus/animations-dbus.h:
```diff
@@ -0,0 +1,22 @@
+
+/* Pull in other header files */
+#include "animations-dbus-errors.h"
+#include "animations-dbus-version.h"
```
Please keep this in alphabetical order if possible

smspillaz replied:
I've done that as much as possible, though in a later commit `#include "animations-dbus-server-types.h"` needed to come before everything else.

cosimoc commented on animations-dbus/animations-dbus-server.c:
```diff
@@ -0,0 +1,624 @@
+  if (!g_hash_table_lookup_extended (priv->animation_manager_ids,
+                                     GUINT_TO_POINTER (animation_manager_id),
+                                     NULL,
+                                     (gpointer *) &client_name))
```
Why do you need to use the extended method?

smspillaz replied:
I think I originally had it when the hash table mapped from ints-to-ints, and we needed to rely on the `gboolean` return value, but that's no longer the case. I'll change it back to the non-extended version.

cosimoc commented on animations-dbus/animations-dbus-server.c:
```diff
@@ -0,0 +1,624 @@
+ *
+ * Get a #GPtrArray of surfaces that this #AnimationsDbusServer is tracking.
+ *
+ * Returns: (transfer container) (element-type AnimationsDbusServerSurface): A #GPtrArray
```
This is transfer none

smspillaz replied:
I think from our discussions that we had around https://github.com/endlessm/libcontentfeed/blob/master/src/feed-model-ordering.c#L397 we worked out that it was `transfer container`. Though I believe @pwithnall knows more about this .. it was a rather esoteric part of the way that transfer rules for container types work in the bindings.

pwithnall replied:
It’s only `transfer container` if you’re transferring a ref to the `GPtrArray`. If you’re just returning a pointer to it without any ownership transfer, it’s `transfer none`.

The confusion before was between `transfer container` and `transfer full`. In both cases, the function was transferring a ref to the `GPtrArray`. In the `transfer full` case, bindings would then unref each of the elements in the `GPtrArray` before unreffing the `GPtrArray` itself — but the `GPtrArray` would then continue to call its `GDestroyNotify` function on each element again, resulting in a double-unref for each element. In the `transfer container` case, the bindings don’t unref each element before calling `g_ptr_array_unref()`.

cosimoc commented on animations-dbus/animations-dbus-server.c:
```diff
@@ -0,0 +1,624 @@
+
+  if (!g_ptr_array_remove_fast (priv->animatable_surfaces, server_surface))
+    {
+      g_set_error (error,
```
I think that passing a surface that is not tracked should be a hard programmer error (e.g. a `g_assert()`), not a recoverable error here.

smspillaz replied:
Perhaps, though I'd rather that the shell didn't crash due to a programmer error if we can avoid it.

cosimoc commented on animations-dbus/animations-dbus-server.c:
```diff
@@ -0,0 +1,624 @@
+       * use an additional hash table here to avoid this cost
+       * but this seems like a simpler solution. */
+      unsigned int n_removed = g_hash_table_foreach_remove (priv->animation_manager_ids,
+                                                            find_matching_value,
```
Use `g_str_equal()`

smspillaz replied:
This also works - fixed.

smspillaz replied:
Ah, damn, that won't work since the calling convention is different:

```
../animations-dbus/animations-dbus-server.c:278:61: warning: passing argument 2 of ‘g_hash_table_foreach_remove’ from incompatible pointer type [-Wincompatible-pointer-types]
                                                             g_str_equal,
```

cosimoc commented on animations-dbus/animations-dbus-server.c:
```diff
@@ -0,0 +1,624 @@
+  if (g_hash_table_lookup_extended (priv->client_name_watches,
+                                    name,
+                                    NULL,
+                                    (gpointer *) &watch_id))
```
Don't you need to use GPOINTER_TO_UINT() here?

smspillaz replied:
I think in the case of `g_hash_table_lookup_extended`, it wants a `gpointer *`. That said, maybe I can use `&GUINT_TO_POINTER (watch_id)`

smspillaz replied:
Damn, `GUINT_TO_POINTER` returns an lvalue:

```
../animations-dbus/animations-dbus-server.c: In function ‘on_animation_manager_owner_name_lost’:
../animations-dbus/animations-dbus-server.c:267:37: error: lvalue required as unary ‘&’ operand
                                     &(GUINT_TO_POINTER (watch_id))))
```

pwithnall replied:
This will have to be:
```
gpointer watch_id_ptr;
if (g_hash_table_lookup_extended (priv->client_name_watches, name, NULL, &watch_id_ptr))
  {
    gint watch_id = GPOINTER_TO_INT (watch_id_ptr);
    …
  }
```

(`GPOINTER_TO_INT` can be more than just a cast on some platforms.)

cosimoc commented on animations-dbus/animations-dbus-server.c:
```diff
@@ -0,0 +1,624 @@
+animations_dbus_server_new_async (AnimationsDbusServerEffectFactory *factory,
+                                  GCancellable                      *cancellable,
+                                  GAsyncReadyCallback                callback,
+                                  gpointer                           user_data)
```
You may be able to pass a `GDBusConnection` to this method to make the async initialization simpler -- clients will almost certainly have access to such an object already.

Also, we should have a `_finish()` method that wraps `g_async_initable_finish()`.

smspillaz replied:
I added a `_finish` method shortly after this review was done (I'll document the other changes too).

I suppose that we could allow clients to pass their own connection ... my thought here was that the server object could exist independently of any other connection. Is it considered bad practice to have multiple connections from the same app?

pwithnall replied:
Generally you’d want to use a single connection to the bus from a single app. The only cases where I’ve ever done differently are:
 * When I’m connecting to the session bus *and* the system bus
 * Where I want to authenticate differently to the same D-Bus service (since authentication is typically based on the client connection’s unique name)

smspillaz replied:
Makes sense. I realized that having it accept a connection would be useful for the tests too, so lets do that.

pwithnall replied:
> I realized that having it accept a connection would be useful for the tests too,

Yes, very much so. You don’t want to tie the tests down to a particular bus instance (definitely not the system bus, ideally not the user’s real session bus either).

cosimoc commented on animations-dbus/animations-dbus-server-surface.c:
```diff
@@ -0,0 +1,691 @@
+  AnimationsDbusServerSurface *server_surface = ANIMATIONS_DBUS_SERVER_SURFACE (object);
+  AnimationsDbusServerSurfacePrivate *priv = animations_dbus_server_surface_get_instance_private (server_surface);
+
+  priv->server = NULL;
```
This does not seem necessary.

smspillaz replied:
It isn't strictly speaking necessary - I just wanted to make it clear that `server` is a non-owning reference (perhaps it may be a good usecase for `GWeakRef`).

pwithnall replied:
No need to use `GWeakRef` unless you need the notification callback about the weak ref being cleared to `NULL` when the object is finalised.

Personally, I think it would be marginally clearer just to have a comment in this `dispose` function:
```
/* Don’t clear priv->server as it’s not owned */
```

smspillaz replied:
Fair enough - I can just have a comment./

cosimoc commented on animations-dbus/animations-dbus-server-surface.c:
```diff
@@ -0,0 +1,691 @@
+      g_variant_dict_insert_value (&vardict,
+                                   event,
+                                   g_variant_new_strv ((const char * const *) names->pdata,
+                                                       names->len));
```
There is no need for a double indirection here... We could use a variant builder instead of the `GPtrArray` to build the strv.

smspillaz replied:
Yeah, using a builder here also works.

smspillaz replied:
Hmm, something strange has started happening since I switched to using GVariantBuilder here:

```
(gnome-shell:65911): GLib-CRITICAL **: g_variant_builder_add_value: assertion '!GVSB(builder)->expected_type || g_variant_is_of_type (value, GVSB(builder)->expected_type)' failed
                                             
Thread 1 "gnome-shell" received signal SIGTRAP, Trace/breakpoint trap.                                                                
_g_log_abort (breakpoint=breakpoint@entry=1) at ../../../../glib/gmessages.c:554
warning: Source file is more recent than executable.
554     }                                        
(gdb) bt
#0  0x00007ffff705c041 in _g_log_abort (breakpoint=breakpoint@entry=1) at ../../../../glib/gmessages.c:554
#1  0x00007ffff705d33d in g_logv (log_domain=0x7ffff709f80e "GLib", log_level=G_LOG_LEVEL_CRITICAL, format=<optimized out>, args=args@entry=0x7fffffffdcd0) at ../../../../glib/gmessages.c:1362
#2  0x00007ffff705d49f in g_log (log_domain=<optimized out>, log_level=<optimized out>, format=<optimized out>) at ../../../../glib/gmessages.c:1403
#3  0x00007fffa75f458e in serialize_attached_effects_to_variant (effects_for_events=0x5555585b7a40) at ../animations-dbus/animations-dbus-server-surface.c:540
#4  0x00007fffa75f47c0 in animations_dbus_server_surface_get_property (object=0x5555585cbb70 [AnimationsDbusServerSurface], prop_id=6, value=0x7fffffffe060, pspec=0x555557389840 [GParamVariant]) at ../animations-dbus/animations-dbus-server-surface.c:592
#5  0x00007ffff733ab51 in object_get_property (value=0x7fffffffe060, pspec=0x555557389840 [GParamVariant], object=0x5555585cbb70 [AnimationsDbusServerSurface]) at ../../../../gobject/gobject.c:1394
#6  0x00007ffff733ab51 in g_object_get_property (object=0x5555585cbb70 [AnimationsDbusServerSurface], property_name=<optimized out>, value=0x7fffffffe060) at ../../../../gobject/gobject.c:2591
```

If I change the `GVariantType` passed to `g_variant_builder_open` to just `"s"`, I get a different assertion:

```
(gnome-shell:65183): GLib-CRITICAL **: g_variant_builder_init: assertion 'g_variant_type_is_container (type)' failed                                         
                                                                                                                                                              
Thread 1 "gnome-shell" received signal SIGTRAP, Trace/breakpoint trap.                                                                                                                                                                                                         
_g_log_abort (breakpoint=breakpoint@entry=1) at ../../../../glib/gmessages.c:554                                                                                                                     
554     ../../../../glib/gmessages.c: No such file or directory.                                                                                                                                
(gdb) bt                                                                                                                                                                                                                                                                        
#0  0x00007ffff705c041 in _g_log_abort (breakpoint=breakpoint@entry=1) at ../../../../glib/gmessages.c:554                                                       
#1  0x00007ffff705d33d in g_logv (log_domain=0x7ffff709f80e "GLib", log_level=G_LOG_LEVEL_CRITICAL, format=<optimized out>, args=args@entry=0x7fffffffdcb0) at ../../../../glib/gmessages.c:1362
#2  0x00007ffff705d49f in g_log (log_domain=<optimized out>, log_level=log_level@entry=G_LOG_LEVEL_CRITICAL, format=format@entry=0x7ffff70a94a4 "%s: assertion '%s' failed") at ../../../../glib/gmessages.c:1403
#3  0x00007ffff705dbb9 in g_return_if_fail_warning (log_domain=<optimized out>, pretty_function=<optimized out>, expression=<optimized out>) at ../../../../glib/gmessages.c:2755                                                                                               
#4  0x00007ffff708fb6e in g_variant_builder_open (builder=0x7fffffffddc0, type=0x7fffa75f7141) at ../../../../glib/gvariant.c:3572                                              
#5  0x00007fffa75f45c4 in serialize_attached_effects_to_variant (effects_for_events=0x5555585e1b00) at ../animations-dbus/animations-dbus-server-surface.c:540
```

It seems that the variant builder wants the opened type to be a subtype of the builder type, but must itself also be a container. If that were the case, how would building a GStrv work?

cosimoc commented on animations-dbus/animations-dbus-server-surface.c:
```diff
@@ -0,0 +1,691 @@
+ * @event: The event to get the highest priority effect on.
+ *
+ * Check the attached effects for @event on @surface and get the highest priority
+ * effect, returning its #AnimationsDbusServerSurfaceAttachedEffect (private data
```
Open paren that never closes

smspillaz replied:
fixed

cosimoc commented on animations-dbus/animations-dbus-server-effect.c:
```diff
@@ -0,0 +1,491 @@
+  GHashTable *ht = g_hash_table_new_full (g_int_hash, g_int_equal, g_free, NULL);
+
+  g_hash_table_insert (ht,
+                       alloc_gtype (G_TYPE_STRING),
```
Do you really need to alloc these? Does GLib not already allocate `GType` structures for its fundamental types when it starts up?

smspillaz replied:
It does, but `g_hash_table` works in terms of `gpointer` which is only guaranteed to be the size of a pointer on the given architecture, whereas `GType` is `guint64` IIRC.

pwithnall replied:
[`GType` is either `gsize` or `gulong`](https://gitlab.gnome.org/GNOME/glib/blob/master/gobject/gtype.h#L375), so you might be able to get away with avoiding the heap allocations, but not portably.

smspillaz replied:
Yeah, portability was the main concern here. This would work on 64bit, but not 32bit.

pwithnall replied:
IIRC, both `gsize` and `gulong` are 32 bits wide on i386 and 64 bits wide on x86_64, so there should be no portability concerns for those two platforms. But I still think that overall it’s less of a headache to just heap allocate the values.

cosimoc commented on animations-dbus/animations-dbus-server-effect.c:
```diff
@@ -0,0 +1,491 @@
+            if (!set_property_from_variant (G_OBJECT (priv->effect_bridge),
+                                            key,
+                                            value,
+                                            &local_error))
```
`local_error` is unused at the moment

smspillaz replied:
It is - we should probably display the error message ;-)

cosimoc commented on animations-dbus/animations-dbus-server-effect.c:
```diff
@@ -0,0 +1,491 @@
+              }
+
+            g_variant_unref (value);
+            g_free (key);
```
Use autoptr if needed.

smspillaz replied:
Yeah, using `autoptr` is probably the best way to go here.

cosimoc commented on animations-dbus/animations-dbus-server-animation-manager.c:
```diff
@@ -0,0 +1,382 @@
+    animations_dbus_server_effect_factory_create_effect (priv->effect_factory,
+                                                         name,
+                                                         settings,
+                                                         error);
```
Isn't this leaked?

smspillaz replied:
It is, yes - thanks for picking up on that.

cosimoc commented on animations-dbus/animations-dbus-client.c:
```diff
@@ -0,0 +1,544 @@
+}
+
+void
+animations_dbus_client_new_async (GCancellable                      *cancellable,
```
I would have clients pass the dbus connection here

smspillaz replied:
See my thoughts on this above - we could potentially do that, but I'm curious to know if owning a separate connection is bad practice.

cosimoc commented on animations-dbus/animations-dbus-client.c:
```diff
@@ -0,0 +1,544 @@
+      if (countdown->error != NULL)
+        {
+          g_task_return_error (task, g_steal_pointer (&countdown->error));
+          return;
```
Remove `return` from both this and the other branch

smspillaz replied:
I rephrased it as an early-return.

cosimoc commented on animations-dbus/animations-dbus-version.h.in:
```diff
@@ -0,0 +1,80 @@
+/* Copyright 2016 Endless Mobile, Inc. */
```
2018

smspillaz commented:
Comments addressed and/or responded to.

I made a few other additions/changes since I moved this to CR Ready, summarized they are:

 - Renamed all the async client API to have an `_async` suffix
 - Added a sync API (no suffixes) since I imagine that's the default usecase. Sync API just wraps the async API with a pushed main context.
 - Fixed up some issues to do with the way that we were using GAsyncInitable
 - Added examples for async and sync client API

TODO list for once the main rounds of review are complete:
 - Look into into a symbol rename `so_that_method_names_arent_so_horribly_long`
 - Update authors

Commit:
tests: Add tests for a client-server scenario

This tests the library through its D-Bus interface, where
we have tests to validate that surfaces advertised on the
server show upon the client, properties propagate properly,
and that effects attached from the client have the expected
effect on the server.

We also test some of the more interesting edge cases, such as
what happens when a client disconnects and server state
can "take over"

https://phabricator.endlessm.com/T23314

Commit:
build: Add basic buildsystem for animations-dbus

Note that this is pretty terrible - it isn't clear to me how
the generated dbus codegen header and source are meant to
be separated - indexing one or the other just gets the same target
which always points to the same file.

Commit:
animations-dbus: Add server side animations-dbus library

The server side library is designed to be used like GNOME-Shell
or other compositors which implement surface level animations.

The server uses animations_dbus_server_new_async() to asynchronously
create a new AnimationsDbusServer object and corresponding
AnimationsDbusConnectionManager which will be exported on the bus. One
of the construct properties that the server must provide is an implementation
of a AnimationsDbusServerEffectFactory which is responsible for constructing
the effect "metaclass" (AnimationsDbusServerEffectBridge) on the server side.

On each surface that the server is managing, the server calls
animations_dbus_server_register_surface along with an
implementation of a AnimationsDbusServerSurfaceBridge, which
is an interface providing properties about the surface that
would be interesting to internal clients. Calling register_surface
will export an AnimationsDbusSurface object on the bus for clients
to interact with.

The AnimationsDbusServerSurfaceBridge implementation should have
an attach_effect method which takes an AnimationsDbusServerEffectBridge
and determines if it can be attached to the surface on that event. If
it can be, attach_effect should return an AnimationsDbusServerSurfaceAttachedEffect
implementation, behind which will be private data for that animation
(for instance, progress, spring state) that exists for that animation
on the given surface. It is the server's responsibility to actually
perform the animation using that private data and the settings for that
animation exist on the attached AnimationsDbusServerEffectBridge.

Clients create and attach animations by first registering themselves
using the RegisterClient method on the exported AnimationsDbusConnectionManager
which takes note of their bus name and returns an object path to an
AnimationsDbusAnimationManager (specific to that client), which they can
call CreateAnimation on to create new animation metaclasses. Clients can
then call the AttachEffect method on the exported AnimationsDbusSurface
to attach their effects to the surface. Once the bus name owned by the
client disappears from the bus, all the client specific state will
be automatically cleaned up (eg, animations are detached, private data
for attached animations is destroyed).

Surfaces can have multiple animations attached to a single event. In
case multiple animations are attached, the highest priority attached one
is selected. If the highest priority animation is destroyed, it is removed
from the list and the next highest priority animation will be used. In
general, the latest attached effect by a client has the highest priority.

Both external clients and the server can register animations with
the server (the server would create a new AnimationManager object
using the animation_dbus_server_create_animation_manager method
on the AnimationsDbusServer). When the server attaches an animation
to an AnimationsDbusServerSurface, it uses
the animations_dbus_server_surface_attach_animation_effect_with_server_priority
which gives that attached effect the lowest priority (for instance, if
an external client has already attached a temporary effect to a surface,
that effect has a higher priority than the one attached by the server). This
allows the server to, for instance, globally attach effects to all surfaces
as the "default", but with the ability for external clients on the bus
to override that configuration temporarily.

https://phabricator.endlessm.com/T23314

smspillaz commented:
I got somewhere with the rest of the feedback and implementing the tests.

For the sake of a backup, I'm just going to put all the pending work in a WIP commit and squash it later.

smspillaz commented:
Update on the TODOs

    Renamed all the async client API to have an _async suffix -> done
    Added a sync API (no suffixes) since I imagine that's the default usecase. Sync API just wraps the async API with a pushed main context. -> done
    Fixed up some issues to do with the way that we were using GAsyncInitable -> done
    Added examples for async and sync client API -> done
    Tests (probably using Jasmine testing against the public API in Gjs or something) -> done
    Update authors -> done

TODO list for once the main rounds of review are complete:

    Look into into a symbol rename so_that_method_names_arent_so_horribly_long

smspillaz commented:
Merging this per discussion with @ptomato and @erikos
